task-launcher/src/tasks/shared/helpers/replayAudio.js

```diff
@@ -1,39 +1,78 @@
-import { jsPsych } from "../../taskSetup";
-import { mediaAssets } from "../../..";
-import { camelize } from "@bdelab/roar-utils";
+import jsPsychAudioMultiResponse from "@jspsych-contrib/plugin-audio-multi-response";
 
+/**
+ * Overrides some of the properties of a jsPsychAudioMultiResponse object to allow for replaying audio.
+ * This function should be called before the trial is added to the timeline.
+ * @param trial the jsPsychAudioMultiResponse object to override
+ * @param jsPsychPluginApi the jsPsych pluginApi from which to retrieve the audio context and audio buffer
+ * @param replayButtonHtmlId the html id of the replay button
+ */
+export function overrideAudioTrialForReplayableAudio(trial, jsPsychPluginApi, replayButtonHtmlId) {
+  if (trial.type !== jsPsychAudioMultiResponse) {
+    throw new Error(`Expected jsPsychAudioTrial to be of type jsPsychAudioMultiResponse but got ${trial.type}`);
+  }
+  if (!jsPsychPluginApi) {
+    throw new Error(`Expected jsPsychPluginApi to be defined but got ${jsPsychPluginApi}`);
+  }
+  if (!replayButtonHtmlId) {
+    throw new Error(`Expected replayButtonHtmlId to be defined but got ${replayButtonHtmlId}`);
+  }
 
-export function setupReplayAudio(audioSource, audioFile) {
-    // Hardcoded since it uses the replayButtonDiv comopnent
-    const replayBtn = document.getElementById('replay-btn');
+  trial.audioReplayOverrides = {
+    originalOnLoad: trial.on_load,
+    originalOnFinish: trial.on_finish,
+    jsPsychPluginApi,
+    promptAudio: trial.stimulus,
+    audioReplaySource: undefined,
+    replayAudioAsyncFunction: async () => {
+      // check whether audio is already playing
+      if (trial.audioReplayOverrides.audioReplaySource) {
+        console.info(`replayAudioAsyncFunction: audio is already playing`);
+        return;
+      }
 
-    if (replayBtn) {
-      // TODO: this only stops the Replay button from being used if it was already used
-        let isAudioPlaying = false;  
-      
-        async function replayAudio() {  
-            if (isAudioPlaying) return
+      const jsPsychAudioCtx = trial.audioReplayOverrides.jsPsychPluginApi.audioContext();
 
-            const jsPsychAudioCtx = jsPsych.pluginAPI.audioContext();
+      const audioAsset = typeof trial.audioReplayOverrides.promptAudio === 'function'
+        ? trial.audioReplayOverrides.promptAudio()
+        : trial.audioReplayOverrides.promptAudio;
+      console.info(`replaying audioId=${audioAsset}`);
+      const audioBuffer = await trial.audioReplayOverrides.jsPsychPluginApi.getAudioBuffer(audioAsset);
 
-            isAudioPlaying = true;
+      const audioSource = jsPsychAudioCtx.createBufferSource();
+      audioSource.buffer = audioBuffer;
```

Comment:
I think from a memory perspective this should be fine because your audioBuffers are being cached, meaning your memory should not go up more than the sum of all the audio buffers.